build/Build.cs

```diff
@@ -99,51 +99,76 @@ class Build : NukeBuild
-                var inputAssemblies = inputFolder.GlobFiles("NewtonSoft.Json.dll", "Octodiff*", "Octopus.*.dll")
+                // CAREFUL: We don't want to expose third-party libraries like Newtonsoft.Json so we definitely want to
+                // internalize those, but we also don't want to hide any Octopus contracts.
+                //
+                // WARNING: There's an apparent bug in il-repack which ignores all types from subsequent assemblies, even
+                // if a set of exclusion regular expressions is provided. To work around this, we do a two-stage merge:
+                // 1) all of the Octopus assemblies into a temporary assembly, with internalization disabled entirely (leaving everything Octopus.* public); and
+                // 2) that temporary assembly plus all of the third-party assemblies, leaving only the types from the first (Octopus temporary) assembly as public.
+                // --andrewh 14/2/2022.
+
+                // Stage 1: Merge all the Octopus assemblies.
+                var stage1Assemblies = inputFolder.GlobFiles("Octopus.*.dll")
                     .Select(x => x.ToString())
-                    .OrderByDescending(x => x.Contains("Octopus.Server.Client.dll"))
+                    .OrderBy(x => x)
+                    .ToArray();
+
+                var temporaryDllPath = inputFolder / "Octopus.Client.ILMerge.Temporary.dll";
+
+                ILRepackTasks.ILRepack(_ => _
+                    .SetAssemblies(stage1Assemblies)
+                    .SetOutput(temporaryDllPath)
+                    .DisableParallel()
+                    .EnableXmldocs()
+                    .SetLib(inputFolder));
+
+                var stage2Assemblies = inputFolder.GlobFiles("Octopus.Client.ILMerge.Temporary.dll",
+                        "NewtonSoft.Json.dll", "Octodiff*")
+                    .Select(x => x.ToString())
+                    .OrderByDescending(x => x.Contains("Octopus.Client.ILMerge.Temporary.dll"))
                     .ThenBy(x => x)
                     .ToArray();
 
+                var outputDllPath = outputFolder / "Octopus.Client.dll";
+
                 ILRepackTasks.ILRepack(_ => _
-                    .SetAssemblies(inputAssemblies)
-                    .SetOutput(outputFolder / "Octopus.Client.dll")
+                    .SetAssemblies(stage2Assemblies)
+                    .SetOutput(outputDllPath)
                     .EnableInternalize()
                     .DisableParallel()
                     .EnableXmldocs()
-                    .SetLib(inputFolder)
-                );
+                    .SetLib(inputFolder));
 
                 DeleteDirectory(inputFolder);
                 MoveDirectory(outputFolder, inputFolder);
             }
         });
 
+    Target Test => _ => _
+        .DependsOn(Compile)
+        .DependsOn(Merge)   // IMPORTANT: Tests must be run _after_ the merge so that we're confident that we're testing the ILMerged code.  -andrewh 14/2/2022.
```

Comment:
Is it possible to assert in test that we are indeed running on the merged code perhaps we could detect that some dll is indeed no loaded.

Reply:
I think there's probably value in that. We already have assertions on whether there are certain .nuspec file references (i.e. there shouldn't be) and the thing still has to link before the tests will run, so we have an _implicit_ test for that, but it would be good to test it explicitly.

Reply:
I'll make a note of that to follow up with in a separate PR given that this issue is currently blocking a customer.